---
title: "BPC Bladder - Survival"
author: "Alex Paynter"
date: "`r Sys.Date()`"
output:
  quarto::quarto_html:
editor_options:
  quarto:
    chunk_output_type: console
format:
  html:
    embed-resources: true
    toc: true
    theme: sandstone 
execute:
  echo: false
  include: false
  warning: false
  message: false
  fig.width: 7
  fig.height: 5
---

```{r}
library(purrr); library(here); library(fs)
purrr::walk(.x = fs::dir_ls(here('R')), .f = source) # also loads lots of packages.
```

```{r, set_gtsummary_theme}
theme_gtsummary_compact()
theme_gtsummary_language("en", big.mark = "") # for now.
```

```{r}
read_wrap_clin <- function(p) {
  read_rds(file = here("data", 'cohort', p))
}

dft_pt <- read_wrap_clin("pt.rds")
dft_ca_ind <- read_wrap_clin("ca_ind.rds")
# we're taking the augmented version, which has TMB columns added.  The existing info is the same.
dft_cpt <- read_wrap_clin("cpt_aug.rds")


```


## Introduction

This document contains analyses related to survival for the GENIE BPC Bladder Cancer Landscape Manuscript group.  The initial report items are borrowed from the prostate cancer group.

The appendix defines (narrowly) a few survival terms.  Looking there may help clarify what I mean in some places, and please ask if not.

## Basics

```{r}
# now in a separate script.
gg_os_dx_stage <- readr::read_rds(
  here('data', 'survival', 'os_dx_by_stage.rds')
)
```


### R3.1.1 Survival from diagnosis

The following plot shows simple Kaplan-Meier estimates from diagnosis for overall survival.  This is broken into groups based on stage at diagnosis.  These groups could be altered, for example Stage IV could be changed to "metastatic at diagnosis".

```{r}
#| include: true

gg_os_dx_stage
```

Notes:

- The counts for At Risk are snapshots in time.  The counts for Censored and Events are cumulative.
- Everyone is either censored or has an event, so the total n for each group is the sum of the last entry for those two.
- The risk set for both groups increases after 0 years due to adjustment for delayed entry.  Participants enter the risk set when they enter the cohort, which can be well after diagnosis.





```{r}
# now in a separate script.
gg_os_dmet_adj <- readr::read_rds(
  here('data', 'survival', 'os_dmet_by_adjustment.rds')
)
```

### R3.1.2 Survival from metastasis

Next we show survival from metastasis.  Metastasis is **not** identical to 'advanced disease' in the bladder cancer cohort because stage IV can be diagnosed with or without mets.  The strategy for identifying metastases in this cohort is detailed in another report.  

This is a good opportunity to show curves which are (recommended) and are not adjusted for left truncation.

```{r}
#| include: true

gg_os_dmet_adj
```





```{r}
dft_trunc_test <- readr::read_rds(
  here('data', 'survival', 'trunc_test.rds')
)
```


### R3.1.3 Testing truncation independence

The following table shows values for the conditional Kendall Tau statistic.  Kendall tau is a measurement of the rank (nonparametric) association between the time to truncation and time to death.  Truncation in our cohort means delayed entry due to NGS being required to join GENIE cohort.  The conditional version of the Tau statistic takes censoring into account, where pairs of observations have to meet certain conditions to be included.  

**Interpretation:** A negative test statistic would indicate that longer times to NGS are associated with shorter survival times than would be expected if the two variables were independent.  A positive test statistic would indicate the opposite.

We repeat this several times for different cohorts.  When we index from metastasis, the difference in indexing time is not important but it *is* important that this only includes participants which had a metastasis at some point.  

```{r}
#| include: true

dft_trunc_test %>%
  filter(method %in% "MB") %>%
  mutate(
    ci_low = estimate - std.error * qnorm(0.975),
    ci_high = estimate + std.error * qnorm(0.975),
    `Cond. Kendall Tau` = cfmisc::est_int_str(estimate, ci_low, ci_high),
    p.value = cfmisc::pval_nejm(p.value)) %>%
  select(` ` = lab, `Cond. Kendall Tau`, p.value) %>%
  flextable(.) %>%
  autofit(.)
```

**Notes:** 

- Based on the results, we might say the dependence seems more for early stage cases than when limit to the most severe cases (metastatic at diagnosis).  However, all appear to have dependence.
- We apply the method of Martin and Betensky (https://doi.org/10.1198/016214504000001538), which should be sufficient for testing the null.  The $\hat{\tau}_{c_2}$ IPW method from https://doi.org/10.1016%2Fj.csda.2013.11.018 was generally similar.
- One hypothesized sequence of events contributing to the negative association between truncation and survival is a worsening cancer which would prompt both testing and increased risk of death.  




## Specific questions


```{r}
gg_first_line_platinum <- readr::read_rds(
  here('data', 'survival', 'first_line_platinum', 'gg_first_line_platinum.rds')
)
```

### R3.2.1 First line platinum chemotherapy

The first exploration we discussed was stating overall survival for first line Carboplatin/Cisplatin combined with Gemcitabine.  By definition (for line of therapy) this only includes participants who were metastatic at the start of the drug combination.  We are not currently including more elaborate regimens such as GemCarbo + Paclitaxel.

```{r}
#| include: true

gg_first_line_platinum
```

**Notes:**

- If we sharpen our question more we can certainly start to add coufounders or do statistical testing.
- Previous GENIE papers have looked at estimates like these to compare to other studies.  These median KM estimates are quite similar to the first study I found addressing this question: https://pubmed.ncbi.nlm.nih.gov/34535437/









```{r}
dft_met_hrd_surv <- readr::read_rds(
  here('data', 'survival', 'hrd_onco', 'met_hrd_surv.rds')
)

dft_met_hrd_surv %<>% 
  remove_trunc_gte_event(
    trunc_var = 'fmr_fcpt_yrs',
    event_var = 'tt_os_first_met_reg_yrs'
  )

dft_met_hrd_surv %<>% mutate(fmr_fcpt_yrs = ifelse(fmr_fcpt_yrs < 0, 0, fmr_fcpt_yrs))

surv_obj_os_fmr <- with(
  dft_met_hrd_surv,
  Surv(
    time = fmr_fcpt_yrs,
    time2 = tt_os_first_met_reg_yrs,
    event = os_first_met_reg_status
  )
)

dft_met_hrd_surv %<>%
  mutate(
    hrd_disp = case_when(
      hrd_before_entry ~ "Oncogenic HRD",
      T ~ "No Onco. HRD"
    )
  )

gg_os_fmr_hrd <- plot_one_survfit(
  dat = dft_met_hrd_surv,
  surv_form = surv_obj_os_fmr ~ hrd_disp,
  plot_title = "OS from first metastatic regimen",
  plot_subtitle = "Adjusted for (independent) delayed entry"
)

  
```




### R3.2.2 HRD pathway

In our Apr 2024 meeting the group expressed an interest in looking at the HRD pathway.  Further details on this general question are borrowed from the prostate cancer group (which had the same interest).

We will look at overall survival comparing those with an oncogenic homologous repair deficiency (HRD) mutation.  We discussed the complication added by lower rates of BRCA2 reported at sites with normal-matched genomic samples, compared with those doing tumor-only sequencing.

These are the restrictions needed to form the cohort for this analysis:

- Include only participants with a metastasis and at least one regimen after that diagnosis.
- For determining HRD covariate:
  - For each subject, include only NGS tests which had a pathology specimen collected before they enter the risk set.  Risk set entry happens when both (1) first NGS test for that subject is reported and (2) the index event, first metastatic regimen, happens.
  - Include only alterations which are oncogenic and include a HUGO symbol in the HRD pathway according to Pearl (2015).   appendix.
  
The following plot shows Kaplan-Meier survival curves for those two groups:

```{r}
#| include: true
gg_os_fmr_hrd
```

```{r}
dft_alt_onco_hrd <- readr::read_rds(
  here('data', 'survival', 'hrd_onco', 'alt_onco_hrd.rds')
)

dft_onco_hrd_before_entry <- dft_alt_onco_hrd %>% 
  filter(dx_path_proc_cpt_yrs < dx_entry) 

dft_onco_hrd_before_entry %<>%
  count(alt_type, hugo, fusion_desc) %>%
  mutate(
    desc = case_when(
      alt_type %in% "Mutation" ~ hugo,
      alt_type %in% "Fusion" ~ fusion_desc,
      T ~ NA_character_
    )
  ) %>%
  select(alt_type, desc, n) %>%
  arrange(alt_type, desc(n))

```

The following table shows the mutations which met the criteria above (oncogenic, HRD pathway, sample taken at the right time).  The total here is greater than the number of subjects in the Oncogenic HRD group above because each person can have more than one HRD alteration.

```{r}
#| include: true
dft_onco_hrd_before_entry %>%
  flextable(.) %>%
  autofit(.) %>% 
  merge_v(j = 1) %>% 
  flextable::valign(valign = "top") %>%
  fix_border_issues(.)
```

*Notes:*

- As expected BRCA2 is prominent here, the tumor-normal vs tumor-only issue may require some thought.

```{r}
dft_hrd_drugs <- dft_met_hrd_surv %>% 
  count(hrd_disp, regimen_drugs) %>%
  group_by(hrd_disp) %>%
  arrange(desc(n)) %>%
  slice(1:5) %>%
  ungroup(.) %>%
  rename(` ` = hrd_disp)
```

Finally, this table shows a list of the top 5 regimens in each HRD group which counted as the first regimen after metastasis.

```{r}
#| include: true
dft_hrd_drugs %>%
  flextable(.) %>%
  autofit(.) %>%
  merge_v(j = 1) %>% 
  flextable::valign(valign = "top") %>%
  fix_border_issues(.)
```

*Notes:*

- Popular drugs seem similar between groups.
- We discussed grouping drugs, hopefully easier with a list.




```{r}
dft_met_hrd_surv_no_adj <- readr::read_rds(
  here('data', 'outcome', 'surv_met_hrd', 'met_hrd_surv.rds')
)

dft_met_hrd_surv_no_adj %<>% mutate(fmr_fcpt_yrs = ifelse(fmr_fcpt_yrs < 0, 0, fmr_fcpt_yrs))

surv_obj_os_fmr_no_adj <- with(
  dft_met_hrd_surv_no_adj,
  Surv(
    time = tt_os_first_met_reg_yrs,
    event = os_first_met_reg_status
  )
)

dft_met_hrd_surv_no_adj %<>%
  mutate(
    hrd_disp = case_when(
      hrd_before_pm_reg ~ "Oncogenic HRD",
      T ~ "No Onco. HRD"
    )
  )

gg_os_fmr_hrd_no_adj <- plot_one_survfit(
  dat = dft_met_hrd_surv_no_adj,
  surv_form = surv_obj_os_fmr_no_adj ~ hrd_disp,
  plot_title = "OS from first metastatic regimen",
  plot_subtitle = "NOT adjusted for delayed entry"
)
```


Finally, we initially discussed running this analysis with an without risk set adjustment (for left truncation).  If we want do this, I think we would have to amend our criterion to say that the NGS sample would need to be collected before the first metastatic regimen (which would now be the time of risk set entry).  There are only `r sum(dft_met_hrd_surv_no_adj$hrd_before_pm_reg)` people who have an HRD mutation before the first met regimen. Here is the KM curve for that analysis:

```{r}
#| include: true
gg_os_fmr_hrd_no_adj
```

*Notes:*

- The difference between HRD and no HRD is not major in either the risk set adjusted graph or this one.
- As we would expect, removing the risk set adjustment changes the survival curves dramatically.  Median survival with adjustment was about 3.5 years for the no-HRD group, >5 here.



### R3.2.2 
