---
title: "BPC Bladder - Survival"
author: "Alex Paynter"
date: "`r Sys.Date()`"
editor_options:
  quarto:
    chunk_output_type: console
format:
  html:
    embed-resources: true
    toc: true
    theme: sandstone 
execute:
  echo: false
  include: false
  warning: false
  message: false
  fig.width: 7
  fig.height: 5
---

```{r}
library(purrr); library(here); library(fs)
purrr::walk(.x = fs::dir_ls(here('R')), .f = source) # also loads lots of packages.
```

```{r, set_gtsummary_theme}
theme_gtsummary_compact()
theme_gtsummary_language("en", big.mark = "") # for now.
```

```{r}
read_wrap_clin <- function(p) {
  read_rds(file = here("data", 'cohort', p))
}

dft_pt <- read_wrap_clin("pt.rds")
dft_ca_ind <- read_wrap_clin("ca_ind.rds")
# we're taking the augmented version, which has TMB columns added.  The existing info is the same.
dft_cpt <- read_wrap_clin("cpt_aug.rds")


```


## Introduction

This document contains analyses related to survival for the GENIE BPC Bladder Cancer Landscape Manuscript group.  The initial report items are borrowed from the prostate cancer group.

The appendix defines (narrowly) a few survival terms.  Looking there may help clarify what I mean in some places, and please ask if not.

## Basics

```{r}
# now in a separate script.
gg_os_dx_stage <- readr::read_rds(
  here('data', 'survival', 'os_dx_by_stage.rds')
)
```


### R3.1.1 Survival from diagnosis

The following plot shows simple Kaplan-Meier estimates from diagnosis for overall survival.  This is broken into groups based on stage at diagnosis.  These groups could be altered, for example Stage IV could be changed to "metastatic at diagnosis".

```{r}
#| include: true

gg_os_dx_stage
```

Notes:

- The counts for At Risk are snapshots in time.  The counts for Censored and Events are cumulative.
- Everyone is either censored or has an event, so the total n for each group is the sum of the last entry for those two.
- The risk set for both groups increases after 0 years due to adjustment for delayed entry.  Participants enter the risk set when they enter the cohort, which can be well after diagnosis.





```{r}
# now in a separate script.
gg_os_dmet_adj <- readr::read_rds(
  here('data', 'survival', 'os_dmet_by_adjustment.rds')
)
```

### R3.1.2 Survival from metastasis

Next we show survival from metastasis.  Metastasis is **not** identical to 'advanced disease' in the bladder cancer cohort because stage IV can be diagnosed with or without mets.  The strategy for identifying metastases in this cohort is detailed in another report.  

This is a good opportunity to show curves which are (recommended) and are not adjusted for left truncation.

```{r}
#| include: true

gg_os_dmet_adj
```





```{r}
dft_trunc_test <- readr::read_rds(
  here('data', 'survival', 'trunc_test.rds')
)
```


### R3.1.3 Testing truncation independence

The following table shows values for the conditional Kendall Tau statistic.  Kendall tau is a measurement of the rank (nonparametric) association between the time to truncation and time to death.  Truncation in our cohort means delayed entry due to NGS being required to join GENIE cohort.  The conditional version of the Tau statistic takes censoring into account, where pairs of observations have to meet certain conditions to be included.  

**Interpretation:** A negative test statistic would indicate that longer times to NGS are associated with shorter survival times than would be expected if the two variables were independent.  A positive test statistic would indicate the opposite.

We repeat this several times for different cohorts.  When we index from metastasis, the difference in indexing time is not important but it *is* important that this only includes participants which had a metastasis at some point.  

```{r}
#| include: true

dft_trunc_test %>%
  filter(method %in% "MB") %>%
  mutate(
    ci_low = estimate - std.error * qnorm(0.975),
    ci_high = estimate + std.error * qnorm(0.975),
    `Cond. Kendall Tau` = cfmisc::est_int_str(estimate, ci_low, ci_high),
    p.value = cfmisc::pval_nejm(p.value)) %>%
  select(` ` = lab, `Cond. Kendall Tau`, p.value) %>%
  flextable(.) %>%
  autofit(.)
```

**Notes:** 

- Based on the results, we might say the dependence seems more for early stage cases than when limit to the most severe cases (metastatic at diagnosis).  However, all appear to have dependence.
- We apply the method of Martin and Betensky (https://doi.org/10.1198/016214504000001538), which should be sufficient for testing the null.  The $\hat{\tau}_{c_2}$ IPW method from https://doi.org/10.1016%2Fj.csda.2013.11.018 was generally similar.
- One hypothesized sequence of events contributing to the negative association between truncation and survival is a worsening cancer which would prompt both testing and increased risk of death.  




## Specific questions

```{r}
dir_surv_plat <- here('data', 'survival', 'first_line_platinum')
gg_first_line_platinum <- readr::read_rds(
  here(dir_surv_plat, 'gg_first_line_platinum.rds')
)
gg_md_ecog_imp_plat <- readr::read_rds(
  here(dir_surv_plat, "gg_cox_imputation_md_ecog.rds")
)
gg_cox_mod_compare_plat <- readr::read_rds(
  here(dir_surv_plat, "gg_cox_mod_compare.rds")
)
```

### R3.2.1 First line platinum chemotherapy

The first exploration we discussed was stating overall survival for first line Carboplatin/Cisplatin combined with Gemcitabine.  By definition (for line of therapy) this only includes participants who were metastatic at the start of the drug combination.  We are not currently including more elaborate regimens such as GemCarbo + Paclitaxel.

```{r}
#| include: true
gg_first_line_platinum
```

**Notes:**

- Previous GENIE papers have looked at estimates like these to compare to other studies.  These median KM estimates are quite similar to the first study that comes up with 12 seconds of Google researh on the topic: https://pubmed.ncbi.nlm.nih.gov/34535437/





#### Models

Unsatisfying, we cannot really answer the question "Is there a survival benefit to carboplatin?" from the analysis above.  This follows because people who take carboplatin may be systematically different from those woh take cisplatin.  We talked especially about the possibility that people who start on cisplatin may be more frail than those who start with carboplatin.

This motivates us to model, and include some potential confounders.  Below I will present three models (all are adjusted for left truncation and right censoring):

1. **Univariate**.  A simple Cox model with one covariate: An indicator for carboplatin use.  Because we have only CarboGem and CisGem regimens in our analysis, this means the reference level is Cisplatin.
2. **Complete Case** A model with the following terms for covariates:
  - `carboplatin` - same as above.
  - `bin_prev_plat` - A binary indicator for previous exposure platinum chemotherapy (that is, before the first line regimen, which comes from the metastatic state).
  - `bin_prev_nonplat` - A binary indicator for previous exposure to any drug other than platinum chemotherapy.
  - `md_ecog_imp_num` - The last ECOG reported in a medical oncologist note before the person enters the risk set.
  - `age_reg_start` - Age when they start their first line chemo regimen.
  - `de_novo_met` - Was this person diagnosed in the metastatic state or some later time?
  - `female` - An indicator for female sex.
  - `institution_*` - Indicator for the treating institution.  Reference is MSK (most common).
  - `race_eth`*` - Indicators for race.
  - **These factors are up for debate.** I just made a list of things that seemed like plausible causal factors on both choice of platinum chemo (cisplatin vs carbo) and survival.  Let me know if there are others things you (as a clinician) think about when deciding on treatments.
3. **Multiple imputation** The same as the complete case model, except that we use multiple imputation so that all participants can be included.  In our model the missing data comes from ECOG scores.  These are (in my opinion) highly valuable, but only about 70% of people have an ECOG score reported before the index time.
  - Our imputation model uses the other variables above (age, sex, etc.) to fill in ECOG scores for people with missing data.  The method is a variant of *predictive mean matching (called MIDAS)* which collects a pool of similar 'donors' (people who are close in age, sex, etc) and uses them to impute ECOG scores.  This is repeated a few times and pooled to factor in our uncertainty here.
  - The payoff for the extra complexity is we **get to include everyone** in the model, while still taking advantage of ECOG scores as a predictor.  
  
The following figure shows what the imputed values look like.  Each dot is one person.  Iteration 0 is the data as-is which has many missing values (not shown).  Each iteration after that is one imputed dataset.  The red dots are imputed values and tbe blue ones are the same non-missing observations.
  
```{r}
#| include: true
#| fig-height: 3
gg_md_ecog_imp_plat
```

*Comments:*

- The distributions of the dots are similar with the imputed and non-imputed values - and we're not imputing anything out of the range of our data.  That's what I'm looking for with diagnostic plots like this.
- Other imputation methods like using regression trees or PMM with a fixed donor pool gave similar results.

The next plot shows the model results from each of the three models above.  A log(HR) less than zero indicates a lower hazard of death, and a log(HR) above zero indicates an increased risk of death.

```{r}
#| include: true
#| fig.height: 5
gg_cox_mod_compare_plat
```

- Those who received carboplatin in this setting have a significantly higher hazard of death than those who received cisplatin, even when adjusting for relevant confounders.  We're closer to answering our question and confirming external results.
- Some of our other hypotheses on strong effects have support as well:  For example, those with previous exposure to platinum chemo have a higher hazard of death, as did female participants and those with de novo metastatic disease.  ECOG scores did not have as strong an effect (though this doesn't mean we don't need to include or impute them).
- One thing to be clear about:  The site numbers are generally not significant, and certainly don't represent the quality of care given at these institutions.  The reason to include these terms is more adjusting for residual confounding we can't model (SES, for example).











```{r}
dft_met_hrd_surv <- readr::read_rds(
  here('data', 'survival', 'hrd_onco', 'met_hrd_surv.rds')
)

dft_met_hrd_surv %<>% 
  remove_trunc_gte_event(
    trunc_var = 'fmr_fcpt_yrs',
    event_var = 'tt_os_first_met_reg_yrs'
  )

dft_met_hrd_surv %<>% mutate(fmr_fcpt_yrs = ifelse(fmr_fcpt_yrs < 0, 0, fmr_fcpt_yrs))

surv_obj_os_fmr <- with(
  dft_met_hrd_surv,
  Surv(
    time = fmr_fcpt_yrs,
    time2 = tt_os_first_met_reg_yrs,
    event = os_first_met_reg_status
  )
)

dft_met_hrd_surv %<>%
  mutate(
    hrd_disp = case_when(
      hrd_before_entry ~ "Oncogenic HRD",
      T ~ "No Onco. HRD"
    )
  )

gg_os_fmr_hrd <- plot_one_survfit(
  dat = dft_met_hrd_surv,
  surv_form = surv_obj_os_fmr ~ hrd_disp,
  plot_title = "OS from first metastatic regimen",
  plot_subtitle = "Adjusted for (independent) delayed entry"
)

  
```




### R3.2.2 HRD pathway

In our Apr 2024 meeting the group expressed an interest in looking at the HRD pathway.  Further details on this general question are borrowed from the prostate cancer group (which had the same interest).

We will look at overall survival comparing those with an oncogenic homologous repair deficiency (HRD) mutation.  We discussed the complication added by lower rates of BRCA2 reported at sites with normal-matched genomic samples, compared with those doing tumor-only sequencing.

These are the restrictions needed to form the cohort for this analysis:

- Include only participants with a metastasis and at least one regimen after that diagnosis.
- For determining HRD covariate:
  - For each subject, include only NGS tests which had a pathology specimen collected before they enter the risk set.  Risk set entry happens when both (1) first NGS test for that subject is reported and (2) the index event, first metastatic regimen, happens.
  - Include only alterations which are oncogenic and include a HUGO symbol in the HRD pathway according to Pearl (2015).   appendix.
  
The following plot shows Kaplan-Meier survival curves for those two groups:

```{r}
#| include: true
gg_os_fmr_hrd
```

```{r}
dft_alt_onco_hrd <- readr::read_rds(
  here('data', 'survival', 'hrd_onco', 'alt_onco_hrd.rds')
)

dft_onco_hrd_before_entry <- dft_alt_onco_hrd %>% 
  filter(dx_path_proc_cpt_yrs < dx_entry) 

dft_onco_hrd_before_entry %<>%
  count(alt_type, hugo, fusion_desc) %>%
  mutate(
    desc = case_when(
      alt_type %in% "Mutation" ~ hugo,
      alt_type %in% "Fusion" ~ fusion_desc,
      T ~ NA_character_
    )
  ) %>%
  select(alt_type, desc, n) %>%
  arrange(alt_type, desc(n))

```

The following table shows the mutations which met the criteria above (oncogenic, HRD pathway, sample taken at the right time).  The total here is greater than the number of subjects in the Oncogenic HRD group above because each person can have more than one HRD alteration.

```{r}
#| include: true
dft_onco_hrd_before_entry %>%
  flextable(.) %>%
  autofit(.) %>% 
  merge_v(j = 1) %>% 
  flextable::valign(valign = "top") %>%
  fix_border_issues(.)
```

*Notes:*

- As expected BRCA2 is prominent here, the tumor-normal vs tumor-only issue may require some thought.

```{r}
dft_hrd_drugs <- dft_met_hrd_surv %>% 
  count(hrd_disp, regimen_drugs) %>%
  group_by(hrd_disp) %>%
  arrange(desc(n)) %>%
  slice(1:5) %>%
  ungroup(.) %>%
  rename(` ` = hrd_disp)
```

Finally, this table shows a list of the top 5 regimens in each HRD group which counted as the first regimen after metastasis.

```{r}
#| include: true
dft_hrd_drugs %>%
  flextable(.) %>%
  autofit(.) %>%
  merge_v(j = 1) %>% 
  flextable::valign(valign = "top") %>%
  fix_border_issues(.)
```

*Notes:*

- Popular drugs seem similar between groups, with investigational drugs being a notable omission in the HRD group.




```{r}
dft_met_hrd_surv_no_adj <- readr::read_rds(
  here('data', 'survival', 'hrd_onco', 'met_hrd_surv.rds')
)

dft_met_hrd_surv_no_adj %<>% mutate(fmr_fcpt_yrs = ifelse(fmr_fcpt_yrs < 0, 0, fmr_fcpt_yrs))

surv_obj_os_fmr_no_adj <- with(
  dft_met_hrd_surv_no_adj,
  Surv(
    time = tt_os_first_met_reg_yrs,
    event = os_first_met_reg_status
  )
)

dft_met_hrd_surv_no_adj %<>%
  mutate(
    hrd_disp = case_when(
      hrd_before_pm_reg ~ "Oncogenic HRD",
      T ~ "No Onco. HRD"
    )
  )

gg_os_fmr_hrd_no_adj <- plot_one_survfit(
  dat = dft_met_hrd_surv_no_adj,
  surv_form = surv_obj_os_fmr_no_adj ~ hrd_disp,
  plot_title = "OS from first metastatic regimen",
  plot_subtitle = "NOT adjusted for delayed entry"
)
```




```{r}
# Read in the data related to this analysis
path_ercc3 <- here('data', 'survival', 'ercc3_plat')
dft_ercc3_alt <- readr::read_rds(here(path_ercc3, 'ercc3_alt.rds'))
gg_surv_ercc3 <- readr::read_rds(here(path_ercc3, 'gg_surv_ercc3.rds'))
dft_surv_plat_ercc3 <- readr::read_rds(here(path_ercc3, 'surv_dat_plat_ercc3.rds'))
```


### R3.2.3 ERCC3, Platinum chemo

The question at hand is "does an ERCC3 alteration confer an advantage to those receiving platinum-based chemotherapy with gemcitabine?"  If yes, it would confirm previous results Neil and others were aware of.

This analysis will unfortunately be limited by sample size in the ERCC3 arm.  We have a grand total of `r nrow(dft_ercc3_alt)` ERCC3 alterations in `r length(unique(dft_ercc3_alt$record_id))` people.  Only `r sum(str_detect(tolower(dft_ercc3_alt$oncogenic), "oncogenic"))` of these alterations are oncogenic, so we did not limit to oncogenic alterations in ERCC3.  When we limit to those who can make it into this survival analysis, we get `r  sum(dft_surv_plat_ercc3$gene_ercc3)` people in the ERCC3 arm.

All the same, we can have a look.  Here are the regimens which qualify as the first platinum chemotherapy:

```{r}
#| include: true
dft_surv_plat_ercc3 %>% count(regimen_drugs) %>%
  flextable(.) %>%
  autofit(.)
```

KM curves for those with and without ERCC3 alterations are as follows:

```{r}
#| include: true
gg_surv_ercc3
```
 
 
```{r}
dft_lot <- readr::read_rds(
  here('data', 'dmet', 'lines_of_therapy', 'lot.rds')
)
```
 
**Notes**

- We can limit to chemo regimens in the metastatic state (better control, lower concern about dependent truncation), but this will likely further reduce the numbers.





```{r}
# Read in the data related to this analysis
path_ercc2 <- here('data', 'survival', 'ercc2_plat')
dft_ercc2_alt <- readr::read_rds(here(path_ercc2, 'ercc2_alt.rds'))
gg_surv_ercc2 <- readr::read_rds(here(path_ercc2, 'gg_surv_ercc2.rds'))
dft_surv_plat_ercc2 <- readr::read_rds(here(path_ercc2, 'surv_dat_plat_ercc2.rds'))
```


### *R3.2.4 ERCC2, Platinum chemo

This is the same as the above analysis, except we substitute ERCC2 for ERCC3.  Requested on Aug 26 meeting.

All the same, we can have a look.  Here are the regimens which qualify as the first platinum chemotherapy:

```{r}
#| include: true
dft_surv_plat_ercc2 %>% count(regimen_drugs) %>%
  flextable(.) %>%
  autofit(.)
```

KM curves for those with and without ERCC2 alterations are as follows:

```{r}
#| include: true
gg_surv_ercc2
```
 
**Notes**

- Notably, this was done after we didn't find anything significant for HRD, DDR or ERCC3. 
- This differs from the analysis of the DDR pathway below in at least three important ways:  (1) different mutations (2) includes carboplatin regimens (3) includes regimens started before or after metastasis in one analysis.
  - This last point is important, because the factor we're analyzing is directly related to the entry time (ERCC3 alteration).



```{r}

dir_fl_imm <- here('data', 'survival', 'first_line_immuno')
dft_fl_io <- readr::read_rds(
  here(dir_fl_imm, 'surv_first_line_immuno.rds')
)

readr::read_rds(
  here(dir_fl_imm, 'gg_os_fl_io.rds')
)
```
 
### R3.2.5 First line immunotherapy

The question here may have been "does first line immunotherapy lead to longer survival than alteratives?", but I want to use this first presentation as a chance to clarify the question.

Here are the drugs used in the first **line of therapy** (following all the rules from our clinical report) which appear to be immunotherapy to me:

```{r}
lot_regex <- paste(
  c(
    'atezolizumab',
    'durvalumab',
    'cemiplimab',
    'nivolumab',
    'pembrolizumab',
    'ipilimumab',
    'tremelimumab'
  ),
  collapse = "|"
)
```


```{r}
#| include: true
dft_lot %>% 
  filter(line_therapy %in% 1) %>% 
  count(regimen_drugs) %>%
  filter(str_detect(tolower(regimen_drugs), lot_regex)) %>%
  arrange(desc(n)) %>%
  flextable(.) %>%
  autofit(.)
```

So there's good news, we probably have the numbers to do an analysis.  With only two immunotherapies used in large numbers, pembrolizumab and atezolizumab, we could choose to try an analysis as "Any immunotherapy vs. None" or separated into Pembro vs. no pembro, atezolizumab vs no atezolizumab.  Narrowing the question will help as we start considering confounders.

**Update:** This was deemed not interesting to pursue further in a June 2023 meeting.






```{r}
dir_met_ddr <- here('data', 'survival', 'ddr_onco')
gg_met_ddr <- readr::read_rds(
  here(dir_met_ddr, 'gg_met_ddr.rds')
)
dft_alt_onco_ddr <- readr::read_rds(
  here(dir_met_ddr, 'alt_onco_ddr.rds')
)
dft_surv_met_ddr <- readr::read_rds(
  here(dir_met_ddr, 'met_ddr_surv.rds')
)
```


```{r}
dft_onco_ddr_before_entry <- dft_alt_onco_ddr %>% 
  filter(dx_path_proc_cpt_yrs < dx_entry) 

dft_onco_ddr_before_entry %<>%
  count(alt_type, hugo, fusion_desc) %>%
  mutate(
    desc = case_when(
      alt_type %in% "Mutation" ~ hugo,
      alt_type %in% "Fusion" ~ fusion_desc,
      T ~ NA_character_
    )
  ) %>%
  select(alt_type, desc, n) %>%
  arrange(alt_type, desc(n))
```


### R3.2.6 DDR pathway

Similar to the HDR pathway results, except we use the DDR pathway genes.  Those genes include the following in our cohort:

```{r}
#| include: true
dft_onco_ddr_before_entry %>%
  flextable(.) %>%
  autofit(.) %>% 
  merge_v(j = 1) %>% 
  flextable::valign(valign = "top") %>%
  fix_border_issues(.)
```

The KM curves for groups with and without a DDR mutation before entering the risk set are:

```{r}
#| include: true
gg_met_ddr
```


```{r}
gg_md_ecog_imp_ddr <- readr::read_rds(
  here(dir_met_ddr, "gg_cox_imputation_md_ecog.rds")
)
gg_cox_mod_compare_ddr <- readr::read_rds(
  here(dir_met_ddr, "gg_cox_mod_compare.rds")
)
```

*Note:* We discussed whether there were enough people with second line GemCis to consider including in this model.  There are 10 people who had second line GemCis in this cohort, and only one of them had an oncogenic DDR mutation.  To me this doesn't seem worth it - adding confusion to the interpretation for that small boost in sample size. 

#### Models

The above KM plots could be confounded.  For example, maybe it's the case that people who have a DDR mutation tend to be less sick than those who do not.  Or possibly it occurs in a group of participants who do tend to do better on Cisplatin (based on race or age, for example).  We are modeling to further investigate and see if the effect magnitude or direction changes when we add in potential confounders (some of these may be precision variables, which is fine).

These are the three models I ran:

- **Univariate** A cox model with only one factor:  A binary variable indicating presence or abscence of DDR mutations at the time the person enters the risk set (starts a metastatic regimen and has had their first genomic test).
- **Complete cases** A model with factors for:
  - DDR alterations (oncogenic) - same as above.
  - Age at regimen start.
  - Sex (reference: male).
  - Race (reference: White).  Ethnicity is not directly modelled so that we can include UHN's data more easily.
  - Institution (reference: DFCI).
  - Last ECOG score - most recent ECOG score (or converted Karnofsky score) from the medical oncologist note data.
- **Multiple imputation** The complete case model above removes any person with incomplete data.  In our model the missing data comes from ECOG scores.  These are (in my opinion) highly valuable, but only about 60% of people have an ECOG score reported before the index time.
  - Our imputation model uses the other variables above (age, sex, etc.) to fill in ECOG scores for people with missing data.  The method is a variant of *predictive mean matching (MIDAS)* which collects a pool of similar 'donors' (people who are close in age, sex, etc) and borrows ECOG scores from them.
  - The payoff for the extra complexity is we get to include everyone in the model, while still taking advantage of ECOG scores as a predictor.
  
The following figure shows what the imputed values look like.  Each dot is one person.  Iteration 0 is the data as-is which has many missing values (not shown).  Each iteration after that is one imputed dataset.  The red dots are imputed values and tbe blue ones are the same non-missing observations.
  
```{r}
#| include: true
#| fig-height: 3
gg_md_ecog_imp_ddr
```

*Comments:*

- The distributions of these dots should be reassuring - the imputed values have about the same proportions as the existing values.
- Other methods like using regression trees or PMM with a fixed donor pool gave similar results.

The next plot shows the model results from each of the three models above.  A log(HR) less than zero indicates a lower hazard of death, and a log(HR) above zero indicates an increased risk of death.

```{r}
#| include: true
#| fig.height: 5
gg_cox_mod_compare_ddr
```

*Comments:*

- The effect for oncogenic DDR mutations remains similar in direction an magnitude with what we saw above:  These people tend to do better than those without an alteration.
- The features are not all on the same scale.  For example, age has a very small confidence bound because it is on the scale of 0-100, rather than 0/1 like most features.
- Some features change dramatically (e.g. unknown/other race and ethnicity) with the multiple imputation model compared to the complete case model.  This makes sense to me: people with unknown/other race or ethnicity tend to have missing ECOG scores, and were excluded from the complete case model.
- Readers of our paper may benefit from seeing at least one model with multiple imputation, since analysis of complete cases is so common in non-RWD work.  Missing data methods are nearly required to leverage ECOG scores. 




```{r}
dft_ddr_neo <- readr::read_rds(
  here('data', 'survival', 'ddr_neoadj', 'ddr_neoadj.rds')
)
gg_ddr_neo_mosaic <- readr::read_rds(
  here('data', 'survival', 'ddr_neoadj', 'gg_neoadj_ddr_mosaic.rds')
)
gg_clin_path_group_stage <- readr::read_rds(
  here('data', 'survival', 'ddr_neoadj', 'gg_clin_path_group_stage.rds')
)


```


### *R3.2.7 DDR neoadjuvant

*Note:* This section was scrapped and rebuilt after our Aug 26, 2024 meeting.

The goal of this analysis is looking at the difference in group stage at the pathological and clinical points among those receiving neoadjuvant therapy.  By requiring that participants have both clinical and pathological staging, we expect to have only people who had an excision (to obtain the pathology sample), but maybe not 100% of the people in that cohort.

The hypothesis is that those with a DDR mutation may receive more benefit from neoadjuvant therapy.  We would expect this to manifest as a higher proportion of people who have a lower group stage at pathological staging than at clinical.  We are taking the DDR mutation information from the first NGS test in each person.  This leaves open a possiblity that they didn't have a DDR mutation at the time they received neoadjuvant therapy, since many of these are long after pathological staging (and neoadjuvant therapy).


```{r}
ft_ajcc_explain <- tribble(
  ~`When T is...`, ~`And N is...`, ~`And M is...`, ~`Then the stage group is...`, ~`Analysis level`,
  "Ta", "N0", "M0", "0a", 1,
  "Tis", "N0", "M0", "0is", 1,
  "T1", "N0", "M0", "I", 1,
  "T2a", "N0", "M0", "II", 2,
  "T2b", "N0", "M0", "II", 2, 
  "T3a,T3b,T4a", "N0", "M0", "IIIA", 3,
  "T1-T4a", "N1", "M0", "IIIA", 4,
  "T1-T4a", "N2,N3", "M0", "IIIB", 5,
  "T4b", "Any N", "M0", "IVA", NA_real_,
  "Any T", "Any N", "M1a", "IVA", NA_real_,
  "Any T", "Any N", "M1b", "IVA", NA_real_,
)

ft_ajcc_explain %<>%
  flextable(.) %>%
  autofit(.) %>%
  bg(i = c(1:3, 6, 8), bg = "#cdb6be") %>%
  bg(i = c(4:5, 7), bg = "#d1b589") %>% 
  bg(i = 9:11, bg = '#a3a3a3') %>%
  merge_v(j = 5) %>%
  align(j = 5, align = 'left')

```


To do this, there's copious processing on the TNM codes supplied in the cancer index dataset.  There are some appendix tables showing those transformations for anyone interested in checking.  Rather than come up with bespoke categories for TNM, we based this on the AJCC manual and clustered the existing categories.  The following table shows the "AJCC Prognostic Stage Groups" table for bladder cancer, and the clusters we'll be using are shaded in colors:

```{r}
#| include: true

ft_ajcc_explain
```

- Stage IV (dark gray) is excluded from this analysis because we don't expect these people to have neoadjuvant therapy by definition.
- Not-so-amusingly this table has some gaps (for example, M1 with no letter, or T0).  We made decisions that I think are reasonable, but again see the appendix.

The upcoming plot is a mosaic plot, which shows the proportion of participants by area.  Keep in mind that the number of people here is relatively small (n = `r nrow(dft_ddr_neo)` overall).

```{r}
#| eval: true
#| include: true
gg_ddr_neo_mosaic
```

*Notes:*

- This shows that a greater proportion of participants with a DDR mutation had no change in stage or a lower stage at path, compared with clin.
- The narrowness of the DDR column shows that we have fewer people in that group (n = `r nrow(filter(dft_ddr_neo, onco_ddr_disp %in% "DDR alt."))`).

It is also useful to look at the actual values for clinical and pathological group stages in these two groups:

```{r}
#| include: true
gg_clin_path_group_stage
```

*Notes:*

- The changes we're seeing appear to be driven by much greater proportions of non-DDR people being classified as IIIA at pathological staging (both N0 and N1).
- There are also several more participants who are stage 0/I in the DDR-altered group at path staging.
- No IVA or IVB is by design - stage IV people were excluded, excuse the legend for the moment.

**Next steps:**

- An appropriate model for this situation is probably a proportional odds model.  We could fit this using our 5 analysis levels over time, showing whether DDR mutations have a propensity to shift more than non-DDR mutated people.
- "Grouped linear" (use the analysis levels above as numbers in the outcome) and a model for change status (improve, same, worsen) would be good checks or backups.
- A sankey or jittered slope plot would be good to show within-person differences over the two time points.















```{r}
dir_line23 <- here('data', 'survival', "line23")
dft_agents_line23 <- readr::read_rds(
  here(dir_line23, 'agent_23_line.rds')
)
gg_line23_legend <- readr::read_rds(
  here(dir_line23, 'gg_line23_legend.rds')
)
gg_line23_taxane <- readr::read_rds(
  here(dir_line23, 'gg_line23_taxane.rds')
)
```




### R3.2.8 Second line therapies    

In August 2024 we discussed looking at second line therapies, specifically with an interest in single agent therapies used in the second line setting, especially taxanes.  Here is a list of all second and third line regimens (limited to those with at least 5 in either line):

```{r}
#| include: true
dft_agents_line23 %>%
  flextable(.) %>% 
  autofit(.)
```
Each person can only have one 2L or 3L regimen, so this is also a count of people.

To me it seems reasonable to look at 2L/3L taxanes based on the numbers we have.  I'm not totally clear about what question we want to pose, so we'll start with an exploratory plot:

```{r}
#| include: true
#| fig-height: 0.75
plot_grid(
    NULL, gg_line23_legend, NULL,
    nrow = 1,
    rel_widths = c(0.4, 0.3, 0.4)
)
```


```{r}
#| include: true
#| fig-height: 6
gg_line23_taxane
```

**Notes:**

- There appears to be very little difference between the two taxanes for third line therapy.  For second line therapy there could be something worth exploring.
- This plot also may be informative to compare to other cohorts without posing any comparative questions.
- Little censoring in this cohort, almost everyone is observed until death.




## *Appendicies

```{r}
dft_tnm_strip_qc <- readr::read_rds(
  here('data', 'cohort', 'tnm_path_clin_stripping_qc.rds')
)

dft_tnm_class_qc <- readr::read_rds(
  here('data', 'cohort', 'tnm_classification_qc.rds')
)

dt_wrapper <- function(x, font.size = '10pt') {
  DT::datatable(
    x,
    style = "bootstrap4",
    fillContainer = F,
    options=list(
      initComplete = htmlwidgets::JS(
        "function(settings, json) {",
        paste0("$(this.api().table().container()).css({'font-size': '", font.size, "'});"),
        "}")
    ) 
  )
}
```


### *TNM classifications

The first step in our group stage data processing is homogenizing the free text.  For example, the strings `T2B`, `Tc2B` and `TcT2b` for `naaccr_clin_t_cd` all clearly mean the same thing.  In this step our goal is to get them all recoded as "2b".  The following table lists every unique case (defined by variable `name` and `before`) and what we translated it into (`after`).  `cases` is just how many times we observed the case.  Blanks in this table represent missing data (NA):

```{r}
#| include: true
dt_wrapper(dft_tnm_strip_qc)
```

Step 2 is taking those refined codes and assigning a stage based on the TNM inputs.  This is done using the AJCC guides, except that we assume a blank M entry is the same as M0 in the logic (as does the AJCC, starting in the 8th edition).  In the table below `comb_*` means the combination of tumor registry and curated data, prioritizing the registry data when available. `group_clust` is the analysis group.  Here are those cases:

```{r}
#| include: true
dt_wrapper(dft_tnm_class_qc)
```

*Note:* While the logic is the same for time = path and time = clin, they are in separate rows to keep the QC purpose of this (I could have made an error in just one).
